src/async_stub.rs

```diff
@@ -0,0 +1,93 @@
+use std::collections::VecDeque;
+use std::io;
+use std::time::{Duration, Instant};
+
+use futures::{
+    channel::mpsc::{self as channel, UnboundedSender},
+    prelude::*,
+    StreamExt,
+};
+
+use crate::{MspPacket, MspPacketParseError, MspParser};
+
+pub struct AsyncMspStub {
+    tx: UnboundedSender<Request>,
+}
+
+struct Request {
+    message: MspPacket,
+    backchannel: UnboundedSender<Result<MspPacket, MspPacketParseError>>,
+}
+
+struct Pending {
+    cmd: u16,
+    backchannel: UnboundedSender<Result<MspPacket, MspPacketParseError>>,
+    ts: Instant,
+}
+
+impl AsyncMspStub {
+    pub fn new<R: AsyncRead + Unpin + Send, W: AsyncWrite + Unpin + Send>(
+        mut reader: R,
+        mut writer: W,
+        timeout: Duration,
+    ) -> io::Result<(Self, impl Future<Output = impl Send> + Send)> {
+        let (tx, mut rx) = channel::unbounded();
+        let f = async move {
+            let mut parser = MspParser::new();
+            let mut buf = [0u8; 1];
+            let mut pending: VecDeque<Pending> = VecDeque::with_capacity(16);
+
+            loop {
+                futures::select! {
+                    // maybe a new response?
+                    _ = reader.read_exact(&mut buf[..]).fuse() =>
+                           match parser.parse(buf[0]){
+                               Ok(Some(response))=>{
+                                while let Some(p) = pending.front_mut(){
```

Comment:
IIUC this code sends msp packet and then awaits for response for that packet, but what if i want to receive packets and answer them. or even ignore the msp ACK?

Reply:
>  what if i want to receive packets

Then this is does not work for you. But likely then you will also not be able to use `libstd` and the async ecosystem, at least that was why I did not focus on it.

> or even ignore the msp ACK?

Then you can just ignore the Timeout response which you will receive if the ACK was never send. I did not focus on this as well, because I went under the assumption that we are always request/response in MSP.

However, recently I've learned quite a lot more about implementing the `Future` trait yourself (without using the `async` keyword). Can you provide me an API proposal? Just a `trait` with all functions which you would like to have. Than I can look into whether I can provide a better implementation, maybe even without the eventloop based on your API proposal.